### PR TITLE
[feat] daVinci-MagiHuman: complete port (4 variants x 2 modes) with bit-exact parity + umbrella HF repo

### PR DESCRIPTION
**Pull request by SolitaryThinker**
## Summary

Ports the full [SII-GAIR + Sand.ai daVinci-MagiHuman](https://github.com/GAIR-NLP/daVinci-MagiHuman) joint
audio-visual generative model family to FastVideo. Covers all four model
variants (base / DMD-2 distilled / SR-540p / SR-1080p) in both T2V and
TI2V (text+image) modes, with bit-exact pipeline-latent parity vs the
official reference and a single umbrella HF repo at
[FastVideo/MagiHuman-Diffusers](https://huggingface.co/FastVideo/MagiHuman-Diffusers).

## Variant matrix (8 entrypoints, all working)

| Variant | T2V | TI2V | DiT | Steps | CFG | Notes |
|---|---|---|---|---|---|---|
| base | yes | yes | base | 32 | 2 | 480x256 mp4 (video + audio) |
| distill | yes | yes | distill | 8 | 1 (no CFG) | DMD-2 distilled, ~4x faster |
| sr_540p | yes | yes | base + sr_540p | 32 + 5 | 2 + cfg-trick | 896x512 (two-stage) |
| sr_1080p | yes | yes | base + sr_1080p | 32 + 5 | 2 + cfg-trick | 1920x1056 (two-stage + block-sparse local-window attention on 32/40 SR DiT layers) |

Architecture notes:
- 15B parameter single-stream transformer (40 layers, hidden=5120, head_dim=128, GQA num_query_groups=8).
- Joint audio-visual denoising in a unified token sequence; no cross-attention.
- SR variants run a separate SR DiT after the base pass, with trilinear-up of the BR latent + ZeroSNR noise + audio mix + per-frame cfg-trick guidance tensor.
- SR-1080p adds block-sparse video->video local-window attention on 32 of 40 SR DiT layers (frame_receptive_field=11), implemented as a 3-block SDPA accumulator that mirrors upstream's FFAHandler / flex_flash_attn_func contract bit-exactly for this layout.

## Parity (14 of 14 local tests bit-exact)

All tests live under \`tests/local_tests/magi_human/\` and run via \`pytest tests/local_tests/magi_human/ -v -s\`.

| Test | Result |
|---|---|
| \`test_magi_human_dit_parity\` (base) | bit-exact (diff_max=diff_mean=0.0) |
| \`test_magi_human_distill_dit_parity\` | bit-exact |
| \`test_magi_human_t5gemma_parity\` | bit-exact |
| \`test_magi_human_sa_audio_parity\` (FV vs Diffusers) | bit-exact |
| \`test_magi_human_sa_audio_official_parity\` (FV vs official) | bit-exact |
| \`test_magi_human_vae_parity\` (Wan VAE decode) | within 8e-4 max (known fp32 op-order drift, OQ-7) |
| \`test_magi_human_pipeline_smoke\` (2 cases) | passes |
| \`test_magi_human_pipeline_latent_parity\` (base T2V) | bit-exact |
| \`test_magi_human_ti2v_pipeline_latent_parity\` | bit-exact |
| \`test_magi_human_sr540p_pipeline_latent_parity[t2v]\` | bit-exact |
| \`test_magi_human_sr540p_pipeline_latent_parity[ti2v]\` | bit-exact |
| \`test_magi_human_sr1080p_pipeline_latent_parity[t2v]\` | bit-exact |
| \`test_magi_human_sr1080p_pipeline_latent_parity[ti2v]\` | bit-exact |

Total: \`14 passed in ~233s\`.

## E2E verification

All 8 (variant x mode) examples runnable via \`python examples/inference/basic/basic_magi_human*.py\`. Latest sequential run:

| Example | Wall-time | Resolution | mp4 size |
|---|---|---|---|
| basic_magi_human.py | 53s | 480x256 | 223 KB |
| basic_magi_human_ti2v.py | 53s | 480x256 | 482 KB |
| basic_magi_human_distill.py | 42s | 480x256 | 135 KB |
| basic_magi_human_distill_ti2v.py | 42s | 480x256 | 378 KB |
| basic_magi_human_sr540p.py | 69s | 896x512 | 479 KB |
| basic_magi_human_sr540p_ti2v.py | 72s | 896x512 | 1.56 MB |
| basic_magi_human_sr1080p.py | 152s | 1920x1056 | 16.82 MB |
| basic_magi_human_sr1080p_ti2v.py | 151s | 1920x1056 | 15.95 MB |

All 8 exit cleanly. Base T2V mp4 hash \`dcf5f2bf6534c7c0d91e7353e42b23db\` is stable across all 43 commits in this branch (no E2E regression at any phase).

## Umbrella HF repo

[FastVideo/MagiHuman-Diffusers](https://huggingface.co/FastVideo/MagiHuman-Diffusers) (public, Apache 2.0):

\`\`\`
FastVideo/MagiHuman-Diffusers/
\u251c\u2500\u2500 README.md (model card with 5 usage snippets, parity table, lazy-load contract)
\u251c\u2500\u2500 base/{model_index.json, transformer/, scheduler/}
\u251c\u2500\u2500 distill/{model_index.json, transformer/, scheduler/}
\u251c\u2500\u2500 sr_540p/{model_index.json, transformer/, sr_transformer/, scheduler/}
\u2514\u2500\u2500 sr_1080p/{model_index.json, transformer/, sr_transformer/, scheduler/}
\`\`\`

User code is now a single HF string per variant:

\`\`\`python
from fastvideo import VideoGenerator
gen = VideoGenerator.from_pretrained(\"FastVideo/MagiHuman-Diffusers/base\")
gen.generate_video(prompt=\"...\", output_path=\"out.mp4\", save_video=True)
\`\`\`

The four cross-variant shared components (Wan 2.2 VAE, T5-Gemma encoder + tokenizer, Stable Audio VAE) are lazy-loaded from their canonical upstream HF repos at first build, so a user running all four variants ends up with ~75 GB total cache instead of ~400 GB.

## Key implementation work this branch landed

Beyond the variant ports, this branch resolved several real production bugs and added cross-cutting infrastructure:

1. **Channel-major video token packing** (commit \`6d190693\`): the pre-Wave-14 \`_img2tokens\` packed video tokens spatial-major (\`(pT pH pW C)\`); upstream \`UnfoldNd\` packs channel-major (\`(C pT pH pW)\`). Fix is a single-character einops change. The bug was completely silent in pipeline parity (both sides used FV's packer) but produced pure noise in production E2E. Wave 14 investigation took multiple bug-hunt + Oracle iterations to localize.

2. **DiT dtype boundary alignment with upstream** (commit \`3a4816cb\`): four cumulative dtype fixes brought DiT parity from \`diff_max=0.5\` to bit-exact:
   - SDPA inputs cast to bf16 (matches upstream \`flash_attn_with_cp\`'s internal cast at \`dit_module.py:508\`)
   - Post-attention upcast to fp32 before per-head gating multiply
   - Drop the residual-stream bf16 cast at the block boundary (keep fp32 between layers)
   - Parity test scheduler single-shift (matches Wave 11 production migration)

3. **Conversion-script \`_FP32_KEEP_SUFFIXES\` fix** (commit \`829f70d3\`): \`--cast-bf16\` was incorrectly downcasting 8 fp32 tensors that the BASE checkpoint (and the FV adapter / final_linear modules) keep as fp32. Distill DiT parity went from \`diff_mean=0.114\` to bit-exact.

4. **Lazy-load infrastructure** (commits \`53ac1985\`, \`e2ef3234\`):
   - \`MagiHumanPipeline.load_modules\` lazy-loads all four shared components from upstream.
   - \`fastvideo/utils.py:maybe_download_model\` recognises \`org/repo/subfolder\` umbrella paths.
   - \`fastvideo/utils.py:verify_model_config_and_directory\` treats \`model_index.json\` as authoritative for which subfolders must exist.

5. **Code-convention cleanup** (commits \`a3c675e4\`, \`690deabd\`, \`78d546ab\`):
   - DiT runs SDPA via FastVideo's \`LocalAttention\` layer (drops manual GQA \`repeat_interleave\`).
   - RoPE uses \`fastvideo.layers.rotary_embedding._apply_rotary_emb\` instead of file-local helpers.
   - \`MagiAttention\` follows Wave 10's loader-owned-dtype pattern except where matching upstream's hardcoded behavior is required for parity (documented at the call site).

## Files changed (43 commits)

Counted at HEAD vs origin/main:
- 13 new files (5 stages, 5 parity tests, 2 examples, 1 helper)
- ~50 modified files
- ~6000 lines added, ~600 removed
- Port-state journal at \`tests/local_tests/magi-human.md\` (~570 lines) documenting the full 14-wave investigation.

## Breaking changes

None. All other model families and shared infrastructure are unchanged. The only shared-infrastructure modifications are:

- \`fastvideo/utils.py:verify_model_config_and_directory\`: relaxed \`vae/\` requirement (now driven by \`model_index.json\` declarations). All other pipelines that bundle vae continue to work.
- \`fastvideo/utils.py:maybe_download_model\`: added 3+ slash umbrella detection. Backwards-compatible (existing 2-component HF ids and local paths unchanged).
- \`fastvideo/models/loader/component_loader.py\`: registers \`sr_transformer\` in the module loaders dict (single-line addition).

## Testing

\`\`\`bash
# Local parity battery (requires CUDA + GAIR/daVinci-MagiHuman shards + converted weights)
pytest tests/local_tests/magi_human/ -v -s

# E2E examples (any of the 8)
python examples/inference/basic/basic_magi_human.py
python examples/inference/basic/basic_magi_human_ti2v.py
python examples/inference/basic/basic_magi_human_distill.py
python examples/inference/basic/basic_magi_human_sr1080p.py
# ... etc
\`\`\`

See \`tests/local_tests/magi-human.md\` for the full setup guide (HF token, conversion, env vars, troubleshooting).

## Citation

\`\`\`bibtex
@article{davinci-magihuman-2026,
  title  = {Speed by Simplicity: A Single-Stream Architecture for Fast Audio-Video Generative Foundation Model},
  author = {SII-GAIR and Sand.ai},
  journal= {arXiv preprint arXiv:2603.21986},
  year   = {2026}
}
\`\`\`

### CODE REVIEW

File: fastvideo/models/dits/magi_human.py [hunk 1]
```diff
@@ -0,0 +1,864 @@
...
+    def _sdpa(self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor) -> torch.Tensor:
+        """Run SDPA on [L, H, D] tensors and return [L, Hq, D]."""
+        if q.numel() == 0:
+            return q.new_empty(q.shape)
+        out = F.scaled_dot_product_attention(
+            q.transpose(0, 1).unsqueeze(0).contiguous(),
+            k.transpose(0, 1).unsqueeze(0).contiguous(),
+            v.transpose(0, 1).unsqueeze(0).contiguous(),
+            enable_gqa=self.cfg.num_heads_q != self.cfg.num_heads_kv,
+        )
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/models/dits/magi_human.py:364]: `enable_gqa` is supported by `torch.nn.functional.scaled_dot_product_attention` since PyTorch 2.5.0 (added Oct 2024). The fastvideo main branch pins torch ≥2.11 (#1277) and magi-human's reference repo targets ≥2.10, both well above the 2.5 threshold. Empirically verified on this branch's pinned env (`torch 2.10.0+cu128`):

```python
>>> q = torch.randn(1, 8, 4, 32); k = torch.randn(1, 2, 4, 32); v = torch.randn(1, 2, 4, 32)
>>> F.scaled_dot_product_attention(q, k, v, enable_gqa=True).shape
torch.Size([1, 8, 4, 32])
```

The DiT parity test (`test_magi_human_dit_parity`) exercises `_sdpa` in production and matches upstream bit-exact (`diff_max=0.000`). False positive — closing.

File: fastvideo/models/dits/magi_human.py [hunk 2]
```diff
@@ -0,0 +1,864 @@
...
+        for frame_idx in range(num_frames):
+            q_start = frame_idx * token_per_frame
+            q_end = q_start + token_per_frame
+            k_start = max(0, (frame_idx - rf) * token_per_frame)
+            k_end = min(num_video_tokens, (frame_idx + rf + 1) * token_per_frame)
+            out[q_start:q_end] = self._sdpa(
+                q_video[q_start:q_end],
+                k_video[k_start:k_end],
+                v_video[k_start:k_end],
+            )
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/models/dits/magi_human.py:419]: Vectorizing the local-window attention is not memory-feasible as a drop-in here. SR-1080p uses `num_frames=101` and `tokens_per_frame ≈ 32 × 72 = 2304`, so the full attention `(N×N)` matrix would be `~232k × 232k ≈ 54B entries` — even at int8 that's ~54 GB of mask alone, which doesn't fit on a single H100. A proper fix is a custom block-sparse attention kernel (FlashAttention-style), which is out of scope for this initial port.

The current per-frame loop bounds the call count to `num_frames × num_layers` SDPA invocations (3232 for SR-1080p), and SDPA itself dominates wall-time per call. Filed as a follow-up perf opportunity, not a correctness issue. Will revisit when we wire MagiAttention's block-sparse kernel into the SR-1080p path.

File: fastvideo/pipelines/basic/magi_human/stages/audio_decoding.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    t = torch.from_numpy(audio.astype(np.float32))
+    new_length = int(t.shape[0] * time_stretching)
+    # t: [L, C] -> [1, C, L] for interpolate
+    t_bct = t.T.unsqueeze(0)
+    t_bct = torch.nn.functional.interpolate(
+        t_bct,
+        size=new_length,
+        mode="linear",
+        align_corners=True,
+    )
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/pipelines/basic/magi_human/stages/audio_decoding.py]: Good catch — fixed in 8f1443f4. The function now calls `scipy.signal.resample` (FFT-based polyphase resampling), which is exactly what upstream `video_process.resample_audio_sinc` uses (verified via `daVinci-MagiHuman/inference/pipeline/video_process.py:25,131` — both import scipy and call `resample(audio, new_length)`). `scipy` is already a direct fastvideo dependency so this avoids a new transitive dep. Function name + docstring now match the implementation.

File: fastvideo/pipelines/basic/magi_human/stages/denoising.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+            # Precompute packed video+audio tokens after any TI2V first-frame
+            # overwrite. Text varies per cond/uncond call and is attached in
+            # _dit_forward via assemble_packed_inputs.
+            static_packed = build_static_packed_inputs(
+                video_latent=video_latent,
+                audio_latent=audio_latent,
+                audio_feat_len=audio_feat_len,
+                patch_size=self.patch_size,
+                coords_style=self.coords_style,
+            )
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/pipelines/basic/magi_human/stages/denoising.py:166]: Good catch — fixed in 4190c720. Added `StaticPackedLayout` + `precompute_static_packed_layout` to `latent_preparation.py`. The latent prep stage now precomputes the step-invariant coords / modality-maps / channel-padding-width once and stashes them on the batch (`batch.magi_static_packed_layout`). Both denoising loops (`denoising.py` + `sr_denoising.py`) pass it back via the new `layout=` arg of `build_static_packed_inputs`, which short-circuits the meshgrid/`torch.full()` work and only rebuilds the per-step `_img2tokens`-derived video/audio token tensors.

Slow path (`layout=None`) is preserved bit-exact for the `build_packed_inputs` one-shot wrapper used in parity tests. Verified slow vs fast equivalence with `torch.equal` across `video_tokens / audio_tokens / video_coords / audio_coords / video_mm / audio_mm / max_ch / video_token_num / audio_feat_len` for both `coords_style="v1"` and `"v2"`. dit / distill_dit / pipeline_smoke / sr540p / sr1080p / vae parity tests all still pass.
